### PR TITLE
Remove boost

### PR DESCRIPTION
**Pull request by flaviojs**
This PR aims to remove all the boost code, one function at a time.

### CODE REVIEW

**Comment by flaviojs** [comment]
Good, I went smoother than expected. :)
I'll cleanup and remake the commits before asking for a review.

**Comment by flaviojs** [comment]
Ok, this is ready for review.

File: src/sgp/FileMan.cc [hunk 1]
```diff
@@ -672,70 +634,72 @@ FILE* FileMan::openForReadingCaseInsensitive(const std::string &folderPath, cons
...
-	std::vector<std::string> paths;
-	boost::filesystem::path path(dirPath);
-	boost::filesystem::directory_iterator end;
-	for(boost::filesystem::directory_iterator it(path); it != end; it++)
+	std::vector<std::string> ret;
+	std::vector<std::string> paths = FindAllFilesInDir(dirPath, sortResults);
+	for (std::string& path : paths)
 	{
-		if(boost::filesystem::is_regular_file(it->status()))
+		RustPointer<char> path_ext(Path_extension(path.c_str()));
```
Comment (selaux) [inline src/sgp/FileMan.cc]: Ideally we should do most of this logic in rust already, but for the time being it think it is ok...

File: src/game/Editor/Sector_Summary.cc [hunk 2]
```diff
@@ -1981,7 +1981,12 @@ static void SummarySaveMapCallback(GUI_BUTTON* btn, INT32 reason)
 		{
 			if( gubOverrideStatus == READONLY )
 			{
-				FileClearAttributes(GCM->getMapPath(gszDisplayName));
+				std::string path = GCM->getMapPath(gszDisplayName);
+				if (!Fs_setReadOnly(path.c_str(), false))
```
Comment (selaux) [inline src/game/Editor/Sector_Summary.cc]: Instincitvely I would say that `FileClearAttributes` does not equal to `Fs_setReadOnly`, but rather the opposite (removing read only flags). But that is just derived from the names of the functions.

File: rust/stracciatella_c_api/src/c/path.rs [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+//! [`std::path`]: https://doc.rust-lang.org/std/path/index.html
+
+use std::ptr;
+
+use crate::c::common::*;
+
+/// Gets the extension of the path.
+/// Returns null if there is no extension.
+#[no_mangle]
+pub extern "C" fn Path_extension(path: *const c_char) -> *mut c_char {
```
Comment (selaux) [inline rust/stracciatella_c_api/src/c/path.rs]: We could also keep `Path` as a type for C++ and convert to `c_char` whenever needed in C++ code. This way we would represent paths as their original OS-dependent representation. I know its not done in other parts of the code, but might be a thought for the future...

File: rust/stracciatella_c_api/src/c/librarydb.rs [hunk 4]
```diff
@@ -32,19 +32,14 @@ pub extern "C" fn LibraryDB_push(
...
-        Err(err) => {
-            remember_rust_error(format!("{:?}", err));
-            false
-        }
-        Ok(_) => {
-            forget_rust_error();
-            true
-        }
+    if let Err(err) = ldb.add_library(data_dir, library) {
+        remember_rust_error(format!("{:?}", err));
```
Comment (selaux) [inline rust/stracciatella_c_api/src/c/librarydb.rs]: Might be good to include data_dir and library in the error message.

File: rust/stracciatella_c_api/src/c/librarydb.rs [hunk 5]
```diff
@@ -79,23 +72,18 @@ pub extern "C" fn LibraryFile_close(file: *mut LibraryFile) {
...
-        Err(err) => {
-            remember_rust_error(format!("{:?}", err));
-            false
-        }
-        Ok(_) => {
-            forget_rust_error();
-            true
-        }
+    if let Err(err) = seek_result {
+        remember_rust_error(format!("{:?}", err));
```
Comment (selaux) [inline rust/stracciatella_c_api/src/c/librarydb.rs]: Might be good to include library file name in the error message.

File: rust/stracciatella_c_api/src/c/librarydb.rs [hunk 6]
```diff
@@ -106,18 +94,13 @@ pub extern "C" fn LibraryFile_read(
...
-        Err(err) => {
-            remember_rust_error(format!("{:?}", err));
-            false
-        }
-        Ok(_) => {
-            forget_rust_error();
-            true
-        }
+    if let Err(err) = file.read_exact(buffer) {
+        remember_rust_error(format!("{:?}", err));
```
Comment (selaux) [inline rust/stracciatella_c_api/src/c/librarydb.rs]: Might be good to include library file name in the error message.

File: rust/stracciatella_c_api/src/c/fs.rs [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+pub extern "C" fn Fs_setReadOnly(path: *const c_char, readonly: bool) -> bool {
+    forget_rust_error();
+    let path = path_from_c_str_or_panic(unsafe_c_str(path));
+    let result = fs::metadata(&path).and_then(|x| {
+        let mut permissions = x.permissions();
+        permissions.set_readonly(readonly);
+        fs::set_permissions(&path, permissions)
+    });
+    if let Err(err) = result {
+        remember_rust_error(format!("Fs_setReadOnly {:?} {}: {}", path, readonly, err));
```
Comment (selaux) [inline rust/stracciatella_c_api/src/c/fs.rs]: I like this format for rust_errors :+1:.

File: rust/stracciatella_c_api/src/c/fs.rs [hunk 8]
```diff
@@ -0,0 +1,209 @@
...
+        }
+        Ok(tempdir) => into_ptr(TempDir { inner: tempdir }),
+    }
+}
+
+/// Checks if the path exists.
+#[no_mangle]
+pub extern "C" fn Fs_exists(path: *const c_char) -> bool {
+    let path = path_from_c_str_or_panic(unsafe_c_str(path));
+    fs::metadata(&path).is_ok()
```
Comment (selaux) [inline rust/stracciatella_c_api/src/c/fs.rs]: You could just use `path.exists()` here.

File: rust/stracciatella_c_api/src/c/fs.rs [hunk 9]
```diff
@@ -0,0 +1,209 @@
...
+use std::time;
+
+use stracciatella::fs;
+
+use crate::c::common::*;
+use crate::c::misc::VecCString;
+
+/// A directory in the filesystem that is automatically deleted.
+/// The contents of the directory are deleted before the directory is deleted.
+pub struct TempDir {
```
Comment (selaux) [inline rust/stracciatella_c_api/src/c/fs.rs]: Should we extract TempDir into its own file and make it independent from `Fs`?  As `Fs_createTempDir` is also not really dependent on `Fs`.

On second thought, is it really necessary do define our own `TempDir`? Can't we just return the one from the crate and define the `TempDir_` functions on that one?

File: rust/stracciatella/src/fs.rs [hunk 10]
```diff
@@ -71,3 +103,73 @@ pub fn resolve_existing_components(path: &Path, base: Option<&Path>, caseless: b
...
+        // use statvfs in unix family
+        // TODO determine which of statvfs64/statfs64/statvfs/statfs are available
+        use std::os::unix::ffi::OsStrExt;
+
+        let mut bytes = path.as_os_str().as_bytes().to_vec();
+        bytes.push(0);
+        let mut data: libc::statvfs = unsafe { std::mem::zeroed() };
+        let result = unsafe { libc::statvfs(bytes.as_ptr() as *const libc::c_char, &mut data) };
+        return match result {
+            0 => Ok(data.f_bfree as u64 * data.f_bsize as u64),
```
Comment (selaux) [inline rust/stracciatella/src/fs.rs]: You could use `u64::from()` in order to get rid of the `#[allow(clippy::cast_lossless)]`, right?

**Comment by selaux** [comment]
I saw one warning, when loading a game. Not sure if it is related:

```
game/Tactical/Tactical_Save.cc: SynchronizeItemTempFile() Reported 0, should be 9
```

File: src/sgp/FileMan.cc (hunk 1 above)
Comment (flaviojs) [reply src/sgp/FileMan.cc]: Yeah, I plan to rethink paths when the non-boost filesystem stuff is moved to rust.

File: src/game/Editor/Sector_Summary.cc (hunk 2 above)
Comment (flaviojs) [reply src/game/Editor/Sector_Summary.cc]: It is "removing the the read only flag" by setting the flag to false.
Maybe it should have a more neutral name? How about `Fs_updateReadOnly` or `Fs_changeReadOnly`?

File: rust/stracciatella_c_api/src/c/fs.rs (hunk 8 above)
Comment (flaviojs) [reply rust/stracciatella_c_api/src/c/fs.rs]: True, but I want to make it clear that it's touching the filesystem.

File: rust/stracciatella_c_api/src/c/fs.rs (hunk 9 above)
Comment (flaviojs) [reply rust/stracciatella_c_api/src/c/fs.rs]: It relies on the filesystem functions to create a special type of directory, so I think it qualifies for `Fs`.

Unfortunately `cbindgen` won't emit an opaque type for the original `TempDir`, so the wrapper is needed.

File: rust/stracciatella_c_api/src/c/path.rs (hunk 3 above)
Comment (flaviojs) [reply rust/stracciatella_c_api/src/c/path.rs]: I plan to rethink paths when the non-boost filesystem stuff is moved to rust.

File: rust/stracciatella/src/fs.rs (hunk 10 above)
Comment (flaviojs) [reply rust/stracciatella/src/fs.rs]: With `u64::from()` you'll get a different warning when the value is already `u64`.
It depends on the target system, in the CI the linux and mac runners had different types.

**Comment by flaviojs** [comment]
That warning is not related, it was around before I started playing with ja2s.
At the time I looked at the source and didn't understand what the warning really meant.
Everything seemed to work fine so I just ignored it.

**Comment by flaviojs** [comment]
I wonder why I didn't implement `fmt::Display` for `LibraryFile` before...

File: rust/stracciatella_c_api/src/c/librarydb.rs (hunk 6 above)
Comment (flaviojs) [reply rust/stracciatella_c_api/src/c/librarydb.rs]: Done

File: rust/stracciatella_c_api/src/c/librarydb.rs (hunk 5 above)
Comment (flaviojs) [reply rust/stracciatella_c_api/src/c/librarydb.rs]: Done

File: rust/stracciatella_c_api/src/c/librarydb.rs (hunk 4 above)
Comment (flaviojs) [reply rust/stracciatella_c_api/src/c/librarydb.rs]: Done

**Comment by selaux** [comment]
Thank you very much :heart:

**Pull request merged**